core/src/main/scala/org/apache/spark/util/collection/ExternalAppendOnlyMap.scala

```diff
@@ -192,12 +193,16 @@ class ExternalAppendOnlyMap[K, V, C](
    * It will be called by TaskMemoryManager when there is not enough memory for the task.
    */
   override protected[this] def forceSpill(): Boolean = {
-    assert(readingIterator != null)
-    val isSpilled = readingIterator.spill()
-    if (isSpilled) {
-      currentMap = null
+    if (isReadingIterator) {
+      assert(readingIterator != null)
+      val isSpilled = readingIterator.spill()
+      if (isSpilled) {
+        currentMap = null
+      }
+      isSpilled
+    } else {
+      false
```

Comment:
I'm not sure I understand how freeing memory would cause GC pressure, unless you're not actually freeing memory.

@samkum did you try this patch without the "spill map on every call" behavior (basically returning "false" instead of spilling the in-memory map when the iterator doesn't exist)?

If that case also results in increased gc, maybe your app just needs more memory. That would be the case where the assert would trigger before.